### ISSUE TITLE
Add type annotation

### ISSUE THREAD

**Issue by raphCode** [google-deepmind/open_spiel#986]
I would like to see type annotations exported from pyspiel, so that code that interacts with pyspiel can be typechecked properly.

Example:
```python
import pyspiel

def foo(state: pyspiel.State) -> bool:
    return state.is_chance_node()
```

`mypy --strict test.py`

```
test.py:1: error: Cannot find implementation or library stub for module named "pyspiel"  [import]
test.py:1: note: See https://mypy.readthedocs.io/en/stable/running_mypy.html#missing-imports
test.py:4: error: Returning Any from function declared to return "bool"  [no-any-return]
Found 2 errors in 1 file (checked 1 source file)
```

While I can silence the import error, code using e.g. `pyspiel.State` can't be checked properly.

**Comment by lanctot** [comment]
Hi @raphCode, yeah that would be nice, but it's a big job and it's not at the top of our priority list at the moment. 

I will bring it up with the team. We don't have the resources to do this ourselves, but it would make a nice community contribution, and I think we'd be open to this if someone started working on it.

**Comment by raphCode** [comment]
> it's a big job

Is it? I would guess the python exposed api is rather small (at least the part I use) and only that needs to be annotated.

If you can provide directions how to add annotations, I can try to start the work for the parts I use.

**Comment by lanctot** [comment]
Oh I assumed you meant the code that sits on top of the core API, because that is exposed from C++ via pybind11: https://github.com/deepmind/open_spiel/blob/master/open_spiel/python/pybind11/pyspiel.cc. I am not sure if pybind11 supports annotations, but if so then yes that would be a lot easier.

**Comment by jblespiau** [comment]
Here is an example within the Tensorflow repository for JAX:

https://github.com/tensorflow/tensorflow/tree/master/tensorflow/compiler/xla/python/xla_extension
This specific example uses Bazel, but it's the same.

TLDR: One should add a file with a `.pyi` extension, with the same name, and it will be used as the pytype definition. I think https://peps.python.org/pep-0484/#stub-files is the PEP defining this.

**Comment by raphCode** [comment]
I tried running mypy's `stubgen` on `lib/python3.10/site-packages/open_spiel`, but it did not produces anything useable: At least I wasn't able to find a definition for `pyspiel.Game` and `pyspiel.State`, which is what I care about.

The actual "meat" of the library seems to be in `lib/python3.10/site-packages/pyspiel.so`, where most other packages have a folder with their import name.
I added a folder `pyspiel` right next to it, an empty file `py.typed` and the following into `pyspiel/__init__.pyi`:
```python
from enum import Enum

class TensorLayout(Enum):
    HWC = ...
    CHW = ...

class State:
    def rewards(self) -> tuple[float]: ...
    def is_chance_node(self) -> bool: ...
    def is_terminal(self) -> bool: ...
    def current_player(self) -> int: ...
    def chance_outcomes(self) -> dict[int, float]: ...
    def observation_tensor(self) -> list[int]: ...
    def legal_actions(self) -> list[int]: ...
    def apply_action_with_legality_check(self, action: int) -> None: ...

class Game:
    def num_players(self) -> int: ...
    def num_distinct_actions(self) -> int: ...
    def max_chance_outcomes(self) -> int: ...
    def new_initial_state(self) -> State: ...
    def observation_tensor_layout(self) -> TensorLayout: ...
    def observation_tensor_shape(self) -> list[int]: ...
    def observation_tensor_size(self) -> int: ...

def load_game(name: str, args: dict[str, float | int | bool] = {}) -> Game: ...
```

With this stub I can successfully typecheck my part of the code that uses openspiel.

Please transform that into a PR, I don't really feel like wanting to understand the python package publishing process, or messing with C++ just to add some files somewhere.

**Comment by raphCode** [comment]
Oh, I got stubgen to run with module runtime introspection, it was not finding my modules previously.
Here is how:

`python -c "from mypy import stubgen; stubgen.main()" -vm pyspiel`

**Comment by lanctot** [comment]
Closing due to inactivity. Please re-open if you're still interested.
